cmd/gcp-controller-manager/app/gke_approver.go

```diff
@@ -600,3 +617,56 @@ func isNotFound(err error) bool {
 	gerr, ok := err.(*googleapi.Error)
 	return ok && gerr.Code == http.StatusNotFound
 }
+
+func ensureNodeMatchesMetadataOrDelete(opts GCPConfig, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest, metricsLabel string, metricsStartTime time.Time, client clientset.Interface) error {
+	// TODO: short-circuit
+
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return nil
+	}
+	nodeName := strings.TrimPrefix(x509cr.Subject.CommonName, "system:node:")
+	if len(nodeName) == 0 {
+		return nil
+	}
+
+	node, err := client.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// if there is no existing Node object, return success
+		return nil
+	}
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return err
+	}
+
+	delete, err := shouldDeleteNode(opts, node)
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return err
+	}
+	if !delete {
+		// if the existing Node object does not need to be removed, return success
+		return nil
+	}
+
+	err = client.CoreV1().Nodes().Delete(nodeName, &metav1.DeleteOptions{Preconditions: metav1.NewUIDPreconditions(string(node.UID))})
+	if apierrors.IsNotFound(err) {
```

Comment:
But we just checked this during `Get`, when would this check be triggered?
Even if we error out here, the approver will retry

Reply:
> But we just checked this during Get, when would this check be triggered?

if the node was removed in between the original Get and the delete attempt

> Even if we error out here, the approver will retry

the goal is to be zero-impact on node bring up time if possible